### PR TITLE
Underlay overhaul [part 1]

### PR DESCRIPTION
**Pull request by marikomedlock**
- Serialization
  - Defined new POJO classes for de/serialization.
  - Serialized the entity and entity groups in separate files, rather than in one huge underlay file.
- Defined new `EntityGroup` classes instead of explicit relationships between entities. Currently there are two types of `EntityGroup`: `CriteriaOccurrence` (e.g. condition - condition_occurrence - person) and `GroupItems` (e.g. brand - ingredient).
- New entry point `Indexer` class generates query inputs needed for the indexing workflows and writes out a bash script, with accompanying SQL query files, to run them. Each `Entity` and `EntityGroup` can generate one or more indexing commands.
- Pull attribute data type by scanning the BQ table schema of the source data.
- Generate UI hints by querying the source data.
- Add option to specify a raw SQL query instead of a `TablePointer` in the underlay. Needed for child-parent relationships for drug ingredient entity.

Coming next in follow-on PRs:
- Instead of a bash script, run the indexing commands from Java directly.
- Switch API endpoints to call new underlay classes.

### CODE REVIEW

File: api/src/main/java/bio/terra/tanagra/underlay/displayhint/EnumVals.java [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+            MAX_ENUM_VALS_FOR_DISPLAY_HINT,
+            valueFieldVar.getAlias());
+        return null;
+      }
+    }
+    return new EnumVals(
+        enumStringVals.stream().map(esv -> new ValueDisplay(esv)).collect(Collectors.toList()));
+  }
+
+  public static EnumVals computeForField(
```
Comment (tjennison-google) [inline api/src/main/java/bio/terra/tanagra/underlay/displayhint/EnumVals.java:100]: Can the two versions of this function be refactored to share code?

File: api/src/main/java/bio/terra/tanagra/underlay/entitygroup/CriteriaOccurrence.java [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+                new Relationship(
+                    OCCURRENCE_TO_PRIMARY_RELATIONSHIP_NAME, occurrenceEntity, primaryEntity));
+    Map<String, AuxiliaryData> auxiliaryData =
+        Map.of(
+            CRITERIA_PRIMARY_ROLLUP_COUNT_AUXILIARY_DATA_NAME,
+            CRITERIA_PRIMARY_ROLLUP_COUNT_AUXILIARY_DATA);
+
+    EntityGroupMapping sourceDataMapping =
+        EntityGroupMapping.fromSerializedForSourceData(
+            serialized.getSourceDataMapping(), dataPointers, relationships, auxiliaryData);
```
Comment (tjennison-google) [inline api/src/main/java/bio/terra/tanagra/underlay/entitygroup/CriteriaOccurrence.java:68]: I don't understand auxiliary data. Why is it specified in a generic fashion like this? How do the "criteriaId" and "primaryCount" strings work? Are they generated attributes that can be requested?

File: api/src/main/java/bio/terra/tanagra/underlay/entitygroup/CriteriaOccurrence.java [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+    return sourceDataMapping.getRelationshipMappings().get(OCCURRENCE_TO_PRIMARY_RELATIONSHIP_NAME);
+  }
+
+  public AuxiliaryDataMapping getCriteriaPrimaryRollupCountAuxiliaryDataMapping() {
+    return indexDataMapping
+        .getAuxiliaryDataMappings()
+        .get(CRITERIA_PRIMARY_ROLLUP_COUNT_AUXILIARY_DATA_NAME);
+  }
+
+  public Query queryCriteriaPrimaryPairs(String criteriaIdAlias, String primaryIdAlias) {
```
Comment (tjennison-google) [inline api/src/main/java/bio/terra/tanagra/underlay/entitygroup/CriteriaOccurrence.java:131]: Do the strings passed here have any purpose? They seem to arbitrary in the one place this is called.

File: api/src/test/resources/config/entity/ConditionOccurrence.json [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+    { "type": "SIMPLE", "name": "stop_reason", "dataType": "STRING" },
+    { "type": "SIMPLE", "name": "visit_occurrence_id", "dataType": "INT64" },
+    { "type": "SIMPLE", "name": "condition_source_value", "dataType": "STRING" },
+    { "type": "SIMPLE", "name": "condition_source_concept_id", "dataType": "INT64" } ],
+
+  "sourceDataMapping": {
+    "dataPointer": "omop_dataset",
+    "tablePointer": { "table": "condition_occurrence" },
+    "attributeMappings": {
+      "id": { "value": { "column": "condition_occurrence_id" } },
```
Comment (tjennison-google) [inline api/src/test/resources/config/entity/ConditionOccurrence.json:19]: Why is `column` needed here? Are there other valid options?

File: pull-credentials.sh [hunk 5]
```diff
@@ -0,0 +1,46 @@
+#!/bin/bash
```
Comment (tjennison-google) [inline pull-credentials.sh:1]: Is this script appropriate for a public repo?

File: .github/workflows/api-test.yaml [hunk 6]
```diff
@@ -42,6 +42,19 @@ jobs:
...
+      - name: Pull credentials
+        id: pull_credentials
+        run: |
+          # For security reasons, Broad prefers we read GHA secrets instead of reading from vault.
+          # this step does the equivalent of the pull-credentials.sh script.
+          # on local machines, the script fetches a SA from Vault.
+          # in GH actions, the SA key is stored in a GH repo secret.
+          # regardless of how it was fetched, tests and scripts expect these
+          # keys to be stored in rendered/
+          mkdir -p rendered/broad/
```
[the remark this answers is not in the document]
Comment (marikomedlock) [reply .github/workflows/api-test.yaml:54]: Yes, we need credentials so that our tests can talk to BQ (e.g. to check if the dataset exists, to fetch data types, to compute display hints). This is a DataBiosphere repo, not a verily-src one, so the tests talk to Broad-owned infrastructure, not Verily-owned. Broad's approach to running tests against their infrastructure is to copy an SA key file into a GH repo secret, and then use those credentials in the GHA.

File: api/src/main/java/bio/terra/tanagra/exception/SystemException.java [hunk 7]
```diff
@@ -0,0 +1,33 @@
+package bio.terra.tanagra.exception;
+
+/**
+ * Custom exception class for system or internal exceptions. These represent errors that the user
```
[the remark this answers is not in the document]
Comment (marikomedlock) [reply api/src/main/java/bio/terra/tanagra/exception/SystemException.java:4]: My thinking here is that there are errors that are understandable to the user (e.g. artifact not found, cohort name uses invalid characters, you don't have access to this underlay) and errors that are not (e.g. BQ is not accepting any requests, our SQL logic got a null ptr exception).

Since I've been overhauling the indexing & query code, I've only hit cases where the details are useless to the user (i.e. Tanagra has a bug, go talk to the admin). I'm hoping that once I get to the artifact storage APIs and implementing access control, a second exception class here will be useful. If it's not, then I'll remove this.

File: api/src/main/java/bio/terra/tanagra/indexing/Indexer.java [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+  public static Indexer fromResourceFile(String underlayResourceFilePath) {
+    return new Indexer(underlayResourceFilePath, READ_RESOURCE_FILE_FUNCTION);
+  }
+
+  public static Indexer fromFile(String underlayFilePath) {
+    return new Indexer(underlayFilePath, READ_FILE_FUNCTION);
+  }
+
+  public void indexUnderlay() throws IOException {
+    // deserialize the POJOs to the internal objects and expand all defaults
```
[the remark this answers is not in the document]
Comment (marikomedlock) [reply api/src/main/java/bio/terra/tanagra/indexing/Indexer.java:49]: OK, I'm happy to use this convention going forward for this repo. I'll reformat comments as I come across them, instead of going through all existing comments.

File: api/src/main/java/bio/terra/tanagra/indexing/WorkflowCommand.java [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+
+public class WorkflowCommand {
```
[the remark this answers is not in the document]
Comment (marikomedlock) [reply api/src/main/java/bio/terra/tanagra/indexing/WorkflowCommand.java:12]: The bash script should go away with what I'm working on now, but I'm using these same classes to do it (i.e. they just have a `run` method now).

RE checking in: This is just [part 1] of the overhaul -- what I demoed to VUMC last week + generating UI hints. I also still have to hook up the API endpoints to the new query classes. I don't think it makes much difference whether I merge [part 1] and then do the rest in separate PRs, or I keep updating this PR. Whatever is easier for you to review is fine with me.

File: api/src/main/java/bio/terra/tanagra/query/CellValue.java [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+
+/**
+ * An interface for the value of a cell within a row within a result table.
+ *
+ * <p>This interface allows us to read data from different databases in a simple but uniform way.
+ * Different database types should implement this for returning values.
+ */
+public interface CellValue {
+  /** Enum for the SQL data types supported by Tanagra. */
+  enum SQLDataType {
```
[the remark this answers is not in the document]
Comment (marikomedlock) [reply api/src/main/java/bio/terra/tanagra/query/CellValue.java:15]: I think it's better to decouple Tanagra's data types from the underlying data storage's data types, so that we can e.g. decide to treat timestamps as strings or dates in Tanagra's API. In the future, we may also want to provide different conversions for different backend data sources.

File: api/src/main/java/bio/terra/tanagra/query/Query.java [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+          ImmutableMap.<String, String>builder()
+              .put("sql", sql)
+              .put("whereSQL", where.renderSQL())
+              .build();
+      sql = StringSubstitutor.replace(template, params);
+    }
+
+    // render each GROUP BY FieldVariable and join them into a single string
+    if (groupBy != null) {
+      // render each GROUP BY FieldVariable and join them into a single string
```
[the remark this answers is not in the document]
Comment (marikomedlock) [reply api/src/main/java/bio/terra/tanagra/query/Query.java:64]: Fixed

File: api/src/main/java/bio/terra/tanagra/query/Query.java [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+          ImmutableMap.<String, String>builder()
+              .put("sql", sql)
+              .put("groupBySQL", groupBySQL)
+              .build();
+      sql = StringSubstitutor.replace(template, params);
+    }
+
+    // render each ORDER BY FieldVariable and join them into a single string
+    if (orderBy != null) {
+      // render each ORDER BY FieldVariable and join them into a single string
```
[the remark this answers is not in the document]
Comment (marikomedlock) [reply api/src/main/java/bio/terra/tanagra/query/Query.java:79]: Fixed

File: api/src/main/java/bio/terra/tanagra/serialization/UFAttribute.java [hunk 13]
```diff
@@ -0,0 +1,85 @@
+package bio.terra.tanagra.serialization;
```
[the remark this answers is not in the document]
Comment (marikomedlock) [reply api/src/main/java/bio/terra/tanagra/serialization/UFAttribute.java:1]: I think Java records looks promising. I'm not sure if it allows inheritance (e.g. each data pointer class extends a base one), though we don't use that too much in the serialization classes so maybe not a dealbreaker.

However, records are available in Java 14+ and Tanagra is currently at Java 11. I want to tackle a version upgrade in a separate PR, since this overhaul [part 1] is already very big. That's why I went with POJOs here.

I think decoupling the user-facing config schema from the internal service/indexing logic is useful. We can be more flexible to requested changes to the config schema, and our internal logic classes are insulated from that.

There is a lot of conversion code, but it's not just applying default values. It's also passing references throughout, to reduce repetition in the user-defined config (e.g. for an attribute that maps to a simple column, the query engine needs the BQ project+dataset id+table name+column name, but we only want the user to specify the column name in the config, and get the rest of the information from higher up). I think we'd need that part regardless.

File: api/src/test/resources/config/entity/ConditionOccurrence.json (hunk 4 above)
Comment (marikomedlock) [reply api/src/test/resources/config/entity/ConditionOccurrence.json:19]: Yes, it could be a foreign-keyed column
```
{
          "column": "vocabulary_id",
          "foreignTable": "vocabulary",
          "foreignKey": "vocabulary_id",
          "foreignColumn": "vocabulary_name" }
```
or include a SQL function wrapper
```
{
          "column": "standard_concept",
          "sqlFunctionWrapper": "(CASE WHEN ${fieldSql} IS NULL THEN 'Source' WHEN ${fieldSql} = 'S' THEN 'Standard' ELSE 'Unknown' END)" }
```

File: api/src/main/java/bio/terra/tanagra/underlay/entitygroup/CriteriaOccurrence.java (hunk 3 above)
Comment (marikomedlock) [reply api/src/main/java/bio/terra/tanagra/underlay/entitygroup/CriteriaOccurrence.java:131]: The reason I included this is because the indexing workflows expect input SQL queries that have specific column names. I'd like to change that, so the caller can override the column names (and in some places you can already) but didn't want to tackle it in this already very big PR. Since Tanagra won't use these specific column names by default, I added a way to alias them to whatever the indexing workflow needs.

File: api/src/main/java/bio/terra/tanagra/underlay/DisplayHint.java [hunk 14]
```diff
@@ -0,0 +1,26 @@
+package bio.terra.tanagra.underlay;
+
+import bio.terra.tanagra.serialization.UFDisplayHint;
+import bio.terra.tanagra.serialization.displayhint.UFEnumVals;
+import bio.terra.tanagra.serialization.displayhint.UFNumericRange;
+import bio.terra.tanagra.underlay.displayhint.EnumVals;
+import bio.terra.tanagra.underlay.displayhint.NumericRange;
+
+public abstract class DisplayHint {
```
[the remark this answers is not in the document]
Comment (marikomedlock) [reply api/src/main/java/bio/terra/tanagra/underlay/DisplayHint.java:9]: No, the underlay will only include entity definition-level hints. Entity instance-level hints like for measurements need to be stored in a DB table.

I haven't thought through how to pass entity instance-level hints to the UI yet. After the overhaul is done, I'll need to figure out how to generate them with an indexing workflow first. Then let's revisit how to return them in query results to the UI.

File: api/src/main/java/bio/terra/tanagra/underlay/Literal.java [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+    this.dataType = DataType.BOOLEAN;
+    this.booleanVal = booleanVal;
+  }
+
+  public static Literal fromSerialized(UFLiteral serialized) {
+    boolean stringValDefined = !Strings.isNullOrEmpty(serialized.getStringVal());
+    boolean int64ValDefined = serialized.getInt64Val() != null;
+    boolean booleanValDefined = serialized.getBooleanVal() != null;
+
+    if (stringValDefined) {
```
[the remark this answers is not in the document]
Comment (marikomedlock) [reply api/src/main/java/bio/terra/tanagra/underlay/Literal.java:40]: Great tip, thanks!

File: api/src/main/java/bio/terra/tanagra/underlay/EntityGroup.java [hunk 16]
```diff
@@ -0,0 +1,136 @@
...
+import java.rmi.RemoteException;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+
+public abstract class EntityGroup {
+  /** Enum for the types of entity groups supported by Tanagra. */
+  public enum Type {
+    ONE_TO_MANY,
```
[the remark this answers is not in the document]
Comment (marikomedlock) [reply api/src/main/java/bio/terra/tanagra/underlay/EntityGroup.java:20]: OK, I'm open to suggestions. Any thoughts? This is specifically for the brand -> ingredient relationship.

File: api/src/main/java/bio/terra/tanagra/underlay/EntityGroup.java [hunk 17]
```diff
@@ -0,0 +1,136 @@
...
+        return OneToMany.fromSerialized(serialized, dataPointers, entities);
+      case CRITERIA_OCCURRENCE:
+        return CriteriaOccurrence.fromSerialized(
+            serialized, dataPointers, entities, primaryEntityName);
+      default:
+        throw new RemoteException("Unknown entity group type: " + serialized.getType());
+    }
+  }
+
+  protected static Entity deserializeEntity(
```
[the remark this answers is not in the document]
Comment (marikomedlock) [reply api/src/main/java/bio/terra/tanagra/underlay/EntityGroup.java:63]: Changed

File: api/src/main/java/bio/terra/tanagra/underlay/datapointer/BigQueryDataset.java [hunk 18]
```diff
@@ -0,0 +1,154 @@
...
+    }
+    if (serialized.getDatasetId() == null || serialized.getDatasetId().isEmpty()) {
+      throw new IllegalArgumentException("No BigQuery dataset ID defined");
+    }
+    if (serialized.getServiceAccountKeyFile() == null
+        || serialized.getServiceAccountKeyFile().isEmpty()) {
+      throw new IllegalArgumentException(
+          "No service account key file defined for BigQuery dataset");
+    }
+    Path serviceAccountKeyFile = Path.of(serialized.getServiceAccountKeyFile());
```
[the remark this answers is not in the document]
Comment (marikomedlock) [reply api/src/main/java/bio/terra/tanagra/underlay/datapointer/BigQueryDataset.java:50]: I don't think this SA key file is going to be the long-term solution, mostly because GCP is recommending people move away from key files. Currently this SA key is used only for indexing. I did this just for expediency, particularly for running the tests.

I don't know whether the indexing/service SAs will typically be different or not. Indexing needs to read source data, write index data; service needs to read both. We should discuss requirements/user-friendly options for pointing Tanagra at credentials (e.g. do we need to handle >1 for a single underlay), maybe in a follow-on PR though.

File: api/src/main/java/bio/terra/tanagra/exception/SystemException.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (tjennison-google) [reply api/src/main/java/bio/terra/tanagra/exception/SystemException.java:4]: I'm not opposed to the idea, but it doesn't appear it's actually serving its purpose here.
1) It's only used in a single place in the PR, but there are are several other types that are used frequently: IOException, UnsupportedOperation, IllegalArgument, ClassCast. That one instance isn't special enough to warrant its own exception. It can be added later if there's actually a need.
2) Remember that users are researchers not programmers. The set of errors they can meaningfully do something about are very limited and will represent a small slice of possible errors. Of the three you listed, the first two are UI bugs outside of rare edge cases so they don't warrant special handling.
3) Errors have to meaningfully transit the API. The third example is useful but the UI needs more information in order to meaningfully handle them. If anything, I would suggest defining a user facing exception instead for specifically cases like that. Consider how Docs displays a special access denied page with a way to request permission. A generic string isn't enough. The response from the API needs a particular response code and machine parsable information that specifies who the owner is or how to request access.

File: api/src/main/java/bio/terra/tanagra/query/CellValue.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (tjennison-google) [reply api/src/main/java/bio/terra/tanagra/query/CellValue.java:15]: Shouldn't this be in the BigQuery implementation of this interface then instead?

File: api/src/main/java/bio/terra/tanagra/serialization/UFAttribute.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (tjennison-google) [reply api/src/main/java/bio/terra/tanagra/serialization/UFAttribute.java:1]: I guess leave it for now but this is the downside to massive PRs like this. If records are the way to go it would make more sense to do the version upgrade first rather than do this poorly and go back to it but I understand that may be painful to do with all of this code to rebase on top.

File: api/src/main/java/bio/terra/tanagra/underlay/EntityGroup.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (tjennison-google) [reply api/src/main/java/bio/terra/tanagra/underlay/EntityGroup.java:20]: As I said, I've referred to them as "groupings". I'm don't remember if I saw a generalized name for concepts in the PR, but perhaps something like `INSTANCE_GROUPING`?

File: api/src/test/resources/config/entity/ConditionOccurrence.json (hunk 4 above)
Comment (tjennison-google) [reply api/src/test/resources/config/entity/ConditionOccurrence.json:19]: Okay, that makes sense.

File: pull-credentials.sh (hunk 5 above)
Comment (tjennison-google) [reply pull-credentials.sh:1]: Sounds good since that's what the other repose do.

File: api/src/main/java/bio/terra/tanagra/underlay/datapointer/BigQueryDataset.java (hunk 18 above)
[the remark this answers is not in the document]
Comment (tjennison-google) [reply api/src/main/java/bio/terra/tanagra/underlay/datapointer/BigQueryDataset.java:50]: Maybe just a TODO for now then?

File: api/src/main/java/bio/terra/tanagra/indexing/Indexer.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (tjennison-google) [reply api/src/main/java/bio/terra/tanagra/indexing/Indexer.java:49]: Sounds good.

File: api/src/main/java/bio/terra/tanagra/serialization/UFAttribute.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (marikomedlock) [reply api/src/main/java/bio/terra/tanagra/serialization/UFAttribute.java:1]: Yeah I actually tried to do the version upgrade first, but it gave me a bunch of errors in the Apache Beam jobs. And given the way the code is structured (1 Gradle project, with the api and workflows in separate sub-projects), we can't upgrade in just the api and leave the workflow unchanged. So I decided to just leave it and focus on the underlay overhaul. I still think we should upgrade, but priority-wise it should come after all our feature work.

File: api/src/main/java/bio/terra/tanagra/underlay/Literal.java [hunk 19]
```diff
@@ -0,0 +1,89 @@
...
+    }
+
+    throw new IllegalArgumentException("No literal values defined");
+  }
+
+  @Override
+  public String renderSQL() {
+    switch (dataType) {
+      case STRING:
+        return "'" + stringVal + "'";
```
[the remark this answers is not in the document]
Comment (marikomedlock) [reply api/src/main/java/bio/terra/tanagra/underlay/Literal.java:64]: Yes, agreed. I think the right way to do this though is to use named parameters in the SQL statement. That way BQ's library will handle the escaping for us. I added a todo here to revisit this, since that's a larger change to pass around a parameter map to all the `renderSQL` methods.

File: api/src/main/java/bio/terra/tanagra/underlay/DisplayHint.java [hunk 20]
```diff
@@ -0,0 +1,26 @@
...
+  public enum Type {
+    ENUM,
+    RANGE
+  }
+
+  public abstract Type getType();
+
+  public UFDisplayHint serialize() {
+    if (getType().equals(Type.ENUM)) {
+      return new UFEnumVals((EnumVals) this);
```
[the remark this answers is not in the document]
Comment (marikomedlock) [reply api/src/main/java/bio/terra/tanagra/underlay/DisplayHint.java:19]: Yep good catch, thanks. Fixed.

File: api/src/main/java/bio/terra/tanagra/serialization/UFDisplayHint.java [hunk 21]
```diff
@@ -0,0 +1,55 @@
...
+      this.type = type;
+      return this;
+    }
+
+    /** Call the private constructor. */
+    public abstract UFDisplayHint build();
+  }
+
+  /** Deserialize to the internal representation of the display hint. */
+  public abstract DisplayHint deserializeToInternal();
```
[the remark this answers is not in the document]
Comment (marikomedlock) [reply api/src/main/java/bio/terra/tanagra/serialization/UFDisplayHint.java:50]: Yes this is a little hard to follow, but I'm not sure how to improve it. I used this same approach for the data pointer classes (see `UF/DataPointer` and `UF/BigQueryDataset`), so the below applies also to that case and any future uses of inheritance.

Both the `serialization` package and the internal `underlay` package contain a base class `UF/DisplayHint` and sub-classes e.g. `UF/NumericRange`. I need to be able to convert in both directions, and I would like to avoid using a `switch(type)` block, since that sort of defeats the purpose of using inheritance. So I used an abstract method on both classes: `serialize` on the internal class, and `deserializeToInternal` on the serialization class.

The deserialization logic includes expanding defaults and pulling information from higher-up objects (e.g. table name, BQ dataset id), so it's significantly more complicated than the serialization logic. I agree with you that it makes more sense to keep the serialized versions as simple JSON mappings and put any non-trivial logic into the internal classes. That's why all the internal objects have a `fromSerialized` method. For sets of classes that use inheritance, there's the additional step of calling the `fromSerialized` method from the serialization class for the reason in the first paragraph.

File: api/src/main/java/bio/terra/tanagra/underlay/EntityGroup.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (marikomedlock) [reply api/src/main/java/bio/terra/tanagra/underlay/EntityGroup.java:20]: Just to clarify, are you suggesting changing `ONE_TO_MANY`->`INSTANCE_GROUPING` or `EntityGroup`->`InstanceGrouping`?

 In the backend, the name for a set of one or more entities that are related in a specific way is `EntityGroup`. There are different types of sets, to support different types of relationships. `CRITERIA_OCCURRENCE` refers to the relationship between a critiera, person, and criteria occurrence entity. `ONE_TO_MANY` refers to the relationship between a brand and ingredient entity (i.e. one brand has many ingredients).

[the remark this answers is not in the document]
Comment (tjennison-google) [reply api/src/main/java/bio/terra/tanagra/underlay/EntityGroup.java:20]: I'm suggesting ONE_TO_MANY to INSTANCE_GROUPING.

File: api/src/main/java/bio/terra/tanagra/underlay/displayhint/EnumVals.java (hunk 1 above)
Comment (marikomedlock) [reply api/src/main/java/bio/terra/tanagra/underlay/displayhint/EnumVals.java:100]: Yes, done.

File: api/src/main/java/bio/terra/tanagra/underlay/entitygroup/CriteriaOccurrence.java (hunk 2 above)
Comment (marikomedlock) [reply api/src/main/java/bio/terra/tanagra/underlay/entitygroup/CriteriaOccurrence.java:68]: - Auxiliary data is basically tables other than the relationship table that we need to execute queries. e.g. static rollup count tables.
- It's specified in a generic fashion so that we don't need to modify the serialization classes when we add a new entity group. I don't know if this is right approach long-term -- an alternative that I actually implemented before changing to this approach, is to have each entity group define its own serialization format. So I'm not wedded to this approach, and think we should revisit it when we add the next entity group.
- "criteriaId" and "primaryCount" strings are just aliases that the indexing workflows expect in the input SQL queries. Similar to the comment below, I think the indexing workflows should be changed to not expect specific column names, but I didn't want to tackle that in this already very big PR.

File: api/src/main/java/bio/terra/tanagra/underlay/EntityGroup.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (marikomedlock) [reply api/src/main/java/bio/terra/tanagra/underlay/EntityGroup.java:20]: Summarizing conversation from stand-up: I think `INSTANCE_GROUPING` is more vague than `ONE_TO_MANY`, but agree we should pick a name here that conveys information about how the entity instances are related. Other  possibilities: `PARENT_CHILD`, `GROUP_ITEMS`, ...? Fine to resolve this in a separate PR if we don't agree on something before this is ready to be merged. Code change is just a simple find/replace.

[the remark this answers is not in the document]
Comment (tjennison-google) [reply api/src/main/java/bio/terra/tanagra/underlay/EntityGroup.java:19]: I would choose GROUP_ITEMS (or maybe ITEM_GROUP?).

[the remark this answers is not in the document]
Comment (marikomedlock) [reply api/src/main/java/bio/terra/tanagra/underlay/EntityGroup.java]: Done.

**Review by tjennison-google — APPROVED** [review]
